### PR TITLE
New Modal2 flow

### PR DESCRIPTION
**Pull request by chrisnojima**
tldr; Use `Modal2` for the 'body' of your modal. Use `navigationOptions.modal2: true` to have the router handle the backdrops / close button etc

This moves the logic of presenting a modal into a custom navigator. This allows the desktop side to more closely act like the mobile side. I applied this to the team building modals as they're likely one of the more complicated examples

In desktop navigationOptions allow you to specify `modal2: true` which opts you into this behavior.

After this we'll start to convert all modals to use this and `common-adapters/modal2` to unify the modal styles (a longtime wish of design)

Additionally:
- [x] Handle if you click and drag the whole app dismissing modals (cc: @cecileboucheron )
- [x] Handle team building side effect which can nav you up even when a modal isn't up

### CODE REVIEW

File: shared/actions/team-building.tsx [hunk 1]
```diff
@@ -8,7 +9,17 @@ import * as RPCTypes from '../constants/types/rpc-gen'
 import {TypedState} from '../constants/reducer'
 import {validateEmailAddress} from '../util/email-address'
 
-const closeTeamBuilding = () => RouteTreeGen.createNavigateUp()
```
Comment (chrisnojima) [inline shared/actions/team-building.tsx:11]: this naively navigated up without any context , now it checks if its actually a modal still

File: shared/actions/team-building.tsx [hunk 2]
```diff
@@ -144,10 +155,34 @@ export function filterForNs<S, A, L, R>(
...
 
+const namespaceToRoute = new Map([
+  ['chat2', 'chatNewChat'],
+  ['crypto', 'cryptoTeamBuilder'],
+  ['teams', 'teamsTeamBuilder'],
+  ['people', 'peopleTeamBuilder'],
+  ['wallets', 'walletTeamBuilder'],
+])
+
+const maybeCancelTeamBuilding = (namespace: TeamBuildingTypes.AllowedNamespace) => (
```
Comment (chrisnojima) [inline shared/actions/team-building.tsx:166]: we cancel if we nav away

File: shared/common-adapters/modal2.tsx [hunk 3]
```diff
@@ -0,0 +1,276 @@
+import * as React from 'react'
+import * as Styles from '../styles'
+import {LayoutChangeEvent} from 'react-native'
+import ScrollView from './scroll-view'
+import {Box2, Box} from './box'
+import BoxGrow from './box-grow'
+import Text from './text'
+import {useTimeout} from './use-timers'
+
+const Kb = {
```
Comment (chrisnojima) [inline shared/common-adapters/modal2.tsx:10]: Basically a copy of Modal but without the inner popup dialog, which is moved into the custom navigator

File: shared/router-v2/router.desktop.tsx [hunk 4]
```diff
@@ -364,10 +470,65 @@ const styles = Styles.styleSheetCreate(
...
+          cursor: 'pointer',
+          padding: Styles.globalMargins.tiny,
+          position: 'absolute',
+          right: Styles.globalMargins.tiny * -4,
+          top: 0,
+        },
+      }),
+      modal2Container: {
+        ...Styles.globalStyles.fillAbsolute,
+        backgroundColor: 'rgba(255,0,0,0.2)', // Styles.globalColors.black_50OrBlack_60,
```
Comment (buoyad) [inline shared/router-v2/router.desktop.tsx:486]: :fire:

File: shared/router-v2/router.desktop.tsx [hunk 5]
```diff
@@ -38,121 +38,219 @@ import OutOfDate from '../app/out-of-date'
...
+    (e: React.MouseEvent) => {
+      const {screenX, screenY, target} = e.nativeEvent
+      if (target !== backgroundRef.current) {
+        return
+      }
+      const delta = Math.abs(screenX - mouseDownX) + Math.abs(screenY - mouseDownY)
+      const dismiss = delta < mouseDistanceThreshold
+      setMouseDownX(mouseResetValue)
+      setMouseDownY(mouseResetValue)
+      if (dismiss) {
```
Comment (buoyad) [inline shared/router-v2/router.desktop.tsx:154]: seems to be something up with this, clicking above or below the modal dismisses, but to the left or right doesn't,

Comment (chrisnojima) [reply shared/router-v2/router.desktop.tsx:154]: ah i changed that recently, will fix tnx

File: shared/common-adapters/modal2.tsx [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+  style?: Styles.StylesCrossPlatform
+}
+
+type FooterProps = {
+  content: React.ReactNode
+  hideBorder?: boolean
+  style?: Styles.StylesCrossPlatform
+}
+
+type Props = {
```
Comment (buoyad) [inline shared/common-adapters/modal2.tsx:36]: allowOverflow, onClose, mobileStyle, noScrollView, scrollViewRef are unused. what's plumbing a callback to `onClose` like with this?

Comment (buoyad) [reply shared/common-adapters/modal2.tsx:36]: (Seems ok to skip the scrollview that was builtin in the other one)

Comment (chrisnojima) [reply shared/common-adapters/modal2.tsx:36]: I think we'd mostly not have an Onclose callback. 98% of the time it just navs up. If you want to know more you can use the nav events hook or watch on nav changed event

Comment (buoyad) [reply shared/common-adapters/modal2.tsx:36]: Some places override with a `clearModals`, and in the add to team wizard we need to do that + clear a bunch of state. Both possible with using the navChanged, but also would probably be easy to plumb through navigation options since those are also just simple action dispatches (don't depend on screen state)
